Silabs siwx91x gpdma driver

smalae opened the pull request:
Enable GPDMA driver for siwx91x device

smalae committed:
drivers: dma: siwx91x: GPDMA clock initialization for siwx91x

Clock driver changes required for initializing the GPDMA clock
for the siwx91x driver

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: dma: siwx91x: DTS changes for siwx91x GPDMA driver

1. Create a YAML file for gpdma node
2. Add GPDMA node in the siwx917.dtsi

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: dma: siwx91x: Add siwx91x GPDMA driver

Implement GPDMA driver for siwx91x device

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: dma: siwx91x: dma_chan_blen test app support

Add siwx917_rb4338a board overlay and config files

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

Martinhoff-maker commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,531 @@
+	struct gpdma_siwx91x_data *data = dev->data;
+	volatile uint32_t channel = find_lsb_set(cfg->reg->GLOBAL.INTERRUPT_REG) - 1;
+	volatile uint32_t channel_int_status = cfg->reg->GLOBAL.INTERRUPT_STAT_REG;
+	uint32_t channel_shift = channel * 4;
+	uint32_t abort_mask = (BIT(0) | BIT(3)) << channel_shift;
+	uint32_t desc_fetch_mask = BIT(1) << channel_shift;
+	uint32_t done_mask = BIT(2) << channel_shift;
```
all theses def are not available in a header file ?

Martinhoff-maker commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,531 @@
+
+	if (channel_int_status & abort_mask) {
+		RSI_GPDMA_AbortChannel(GPDMA_HANDLE, channel);
+		cfg->reg->GLOBAL.INTERRUPT_STAT_REG = abort_mask;
```
for all reg operation, please use sys_read(...) and sys_write(...)

smalae replied:
Done

Martinhoff-maker commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,531 @@
+					       data->chan_info[channel].desc_count);
+		data->chan_info[channel].chan_base_desc = NULL;
+		data->chan_info[channel].desc_count = 0;
+		cfg->reg->GLOBAL.INTERRUPT_STAT_REG = done_mask;
```
don't you only want to clean the interrupt for the channel ?

Martinhoff-maker commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,531 @@
+	return 0;
+}
+
+static void siwx91x_gpdma_isr(const struct device *dev)
```
how do you handle it if you have multiple channel isr ?

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,531 @@
+	static ATOMIC_DEFINE(dma_channels_atomic_##inst,                                           \
+			     DT_INST_PROP(inst, silabs_dma_channel_count));                        \
+	SYS_MEM_BLOCKS_DEFINE_STATIC(desc_pool_##inst, 32,                                         \
+				     CONFIG_GPDMA_SILABS_SIWX91X_DESCRIPTOR_COUNT, 512);           \
```
Most tests in WiseConnect align descriptor arrays to 512 or 1024 bytes. However, I changed the alignment value to 32 bytes and ran the tests, which appear to work correctly. The alignment has now been updated to 32 bytes.

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,531 @@
+			xfer_cfg->chnlCtrlConfig.srcFifoMode = 1;
+		}
+
+		if (sys_mem_blocks_alloc_contiguous(data->gpdma_desc_pool, 1, &gpdma_desc_addr)) {
```
I implemented a spin lock and updated the descriptor type for 'gpdma_desc_addr'. I will use mem_block as it handles the allocation/deallocation of memory blocks better than statically allocated ones.

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,531 @@
+#define SIWX91X_GPDMA_INIT(inst)                                                                   \
+	static ATOMIC_DEFINE(dma_channels_atomic_##inst,                                           \
+			     DT_INST_PROP(inst, silabs_dma_channel_count));                        \
+	SYS_MEM_BLOCKS_DEFINE_STATIC(desc_pool_##inst, 32,                                         \
```
Each descriptor block address must be 32-byte aligned for GPDMA. Since sizeof(RSI_GPDMA_DESC_T) is 20 bytes, without 32-byte alignment for each block, some test cases are failing.

jhedberg commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+			xfer_cfg->chnlCtrlConfig.srcFifoMode = 1;
+		}
+
+		key = k_spin_lock(&data->mem_blocks_lock);
+
+		if (sys_mem_blocks_alloc_contiguous(data->gpdma_desc_pool, 1,
+						    (void **)&gpdma_desc_addr)) {
+			if (pPrevDesc == NULL) {
+				return -EINVAL;
+			}
```
This looks like a bug to me. You're returning from the function before unlocking the spin lock.

smalae replied:
Updated

jhedberg commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+	uint16_t max_xfer_size = GPDMA_DESC_MAX_TRANSFER_SIZE - config->source_data_size;
+	const struct dma_block_config *block_addr = config->head_block;
+	RSI_GPDMA_DESC_T *gpdma_desc_addr = NULL;
+	RSI_GPDMA_DESC_T *pPrevDesc = NULL;
```
Is there a good reason to use CamelCase and Hungarian notation for this variable name? (which violates the Zephyr coding style)

smalae replied:
No, it is updated now

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+
+		memset(gpdma_desc_addr, 0, 32);
+
+		if (RSI_GPDMA_BuildDescriptors(GPDMA_HANDLE, xfer_cfg, gpdma_desc_addr,
```
Done

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+			goto free_desc;
+		}
+
+		k_spin_unlock(&data->mem_blocks_lock, key);
```
Updated

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+				     const struct dma_config *config, RSI_GPDMA_DESC_T *xfer_cfg,
+				     uint32_t channel)
+{
+	uint16_t max_xfer_size = GPDMA_DESC_MAX_TRANSFER_SIZE - config->source_data_size;
```
Yes, source_data_size is validated using siwx91x_gpdma_data_size() and it is always less than GPDMA_DESC_MAX_TRANSFER_SIZE

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+	struct dma_context dma_ctx;
+	GPDMA_DATACONTEXT_T gpdma_ctx;
+	struct sys_mem_blocks *gpdma_desc_pool;
+	struct k_spinlock mem_blocks_lock;
```
Updated

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+	dma_callback_t dma_callback;
+	void *cb_data;
+	RSI_GPDMA_DESC_T *chan_base_desc;
+	uint8_t desc_count;
```
Updated

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+	uint8_t reload_compatible;
+};
+
+static bool siwx91x_gpdma_is_priority_valid(uint32_t channel_priority)
```
Done

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+#include "rsi_gpdma.h"
+#include "rsi_rom_gpdma.h"
+
+#define DT_DRV_COMPAT                silabs_siwx91x_gpdma
```
Updated

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+	TRANSFER_MEM_TO_MEM,
+	TRANSFER_MEM_TO_PER,
+	TRANSFER_PER_TO_MEM,
+	TRANSFER_DIR_INVALID,
```
Added a prefix to the symbols and set the values

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,545 @@
+
+		memset(cur_desc, 0, 32);
+
+		ret = RSI_GPDMA_BuildDescriptors((RSI_GPDMA_HANDLE_T)&data->hal_ctx, xfer_cfg,
```
Done

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,545 @@
+
+	if (size > (GPDMA_DESC_MAX_TRANSFER_SIZE - data_size)) {
+		LOG_ERR("Maximum xfer size should be <= %d\n",
+			(GPDMA_DESC_MAX_TRANSFER_SIZE - data_size));
```
Updated

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,545 @@
+
+	stat->busy = sys_read32((mem_addr_t)&cfg->reg->GLOBAL.DMA_CHNL_ENABLE_REG) & BIT(channel)
+			     ? true
+			     : false;
```
done

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,545 @@
+	int ret;
+
+	gpdma_handle = RSI_GPDMA_Init((void *)&data->hal_ctx, &gpdma_init);
+	if (gpdma_handle == NULL) {
```
True, gpdma_handle would be same as &data->hal_ctx. Only way to avoid this is to ignore the return value and check &data->hal_ctx against NULL

smalae commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,545 @@
+		.dma_ctx.dma_channels = DT_INST_PROP(inst, silabs_dma_channel_count),              \
+		.dma_ctx.atomic = dma_channels_atomic_##inst,                                      \
+		.desc_pool = &desc_pool_##inst,                                                    \
+		.chan_info = chan_info_##inst,                                                     \
```
done

smalae committed:
drivers: dma: siwx91x: Sort CMakeLists.txt items

sort library sources and compile time definitions
in wiseconnect CMakeLists.txt

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

jerome-pouiller commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+
+static void siwx91x_gpdma_free_desc(sys_mem_blocks_t *mem_block, void *block)
+{
+	RSI_GPDMA_DESC_T *cur_desc = (RSI_GPDMA_DESC_T *)block;
```
`siwx91x_gpdma_free_desc()` is always called with a `data->chan_info[channel].desc)` as parameter. You can change the type of `block` in a `RSI_GPDMA_DESC_T *`, remove all the useless casts and rename `block` in `cur_desc`.

jerome-pouiller commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,540 @@
+	for (int i = 0; i < config->block_count; i++) {
+		if (block_addr->block_size > max_xfer_size) {
+			LOG_ERR("Maximum xfer size should be <= %d\n", max_xfer_size);
+			return -EINVAL;
```
What about the already allocated blocks?

jhedberg commented on drivers/dma/dma_silabs_siwx91x_gpdma.c:
```diff
@@ -0,0 +1,538 @@
+#include "rsi_gpdma.h"
+#include "rsi_rom_gpdma.h"
+
+#define DT_DRV_COMPAT                silabs_gpdma
```
Is this compatible name really correct, i.e. is the plan to use the same binding also for EFR32?

jhedberg replied:
So you don't want Si91x anywhere in the name, e.g. `silabs,siwx91x-gpdma`? I'm not saying the current name is necessarily wrong, but just trying to understand our philosophy with the binding names.
